File: android/app/src/main/java/com/rbdarts/feature/auth/AuthProvider.kt
package com.rbdarts.feature.auth

import java.time.Instant

enum class AuthProviderId { Google, Facebook, Apple }

data class AuthSession(
    val userId: String,
    val displayName: String,
    val providerId: AuthProviderId,
    val expiresAt: Instant?
)

sealed class AuthProviderException(message: String) : Exception(message) {
    data object Cancelled : AuthProviderException("Sign-in was cancelled.")
    data object ProviderUnavailable : AuthProviderException("The requested provider is unavailable.")
    data object TokenExchangeFailed : AuthProviderException("Unable to exchange provider token.")
}

interface AuthProvider {
    val providerId: AuthProviderId
    suspend fun signIn(): AuthSession
    suspend fun signOut()
}

class AuthProviderRegistry(providers: List<AuthProvider>) {
    private val providersById = providers.associateBy { it.providerId }

    fun provider(providerId: AuthProviderId): AuthProvider =
        providersById[providerId] ?: throw AuthProviderException.ProviderUnavailable
}

File: ios/RBDarts/Features/Auth/AuthProvider.swift
import Foundation

public enum AuthProviderID: String, Codable, Sendable {
    case google
    case facebook
    case apple
}

public struct AuthSession: Equatable, Sendable {
    public let userId: String
    public let displayName: String
    public let providerId: AuthProviderID
    public let expiresAt: Date?
}

public enum AuthProviderError: Error, Equatable {
    case cancelled
    case providerUnavailable
    case tokenExchangeFailed
}

public protocol AuthProviding: Sendable {
    var providerId: AuthProviderID { get }
    func signIn() async throws -> AuthSession
    func signOut() async throws
}

public struct AuthProviderRegistry: Sendable {
    public let providers: [AuthProviderID: AuthProviding]

    public init(providers: [AuthProviding]) {
        self.providers = Dictionary(uniqueKeysWithValues: providers.map { ($0.providerId, $0) })
    }

    public func provider(_ providerId: AuthProviderID) throws -> AuthProviding {
        guard let provider = providers[providerId] else {
            throw AuthProviderError.providerUnavailable
        }
        return provider
    }
}
